sdk/storage/azure-storage-common/src/main/java/com/azure/storage/common/Utility.java

```diff
@@ -216,7 +215,49 @@ public static OffsetDateTime parseDate(String dateString) {
         }
 
         DateTimeFormatter formatter = DateTimeFormatter.ofPattern(pattern, Locale.ROOT);
-        return LocalDateTime.parse(dateString, formatter).atZone(ZoneOffset.UTC).toOffsetDateTime();
+        return new TimeAndFormat(LocalDateTime.parse(dateString, formatter).atZone(ZoneOffset.UTC).toOffsetDateTime(),
+            formatter);
+    }
+
+    /**
+     * Given a String representing a date in a form of the ISO8601 pattern, generates a Date representing it with up to
+     * millisecond precision.
+     *
+     * @param dateString the {@code String} to be interpreted as a <code>Date</code>
+     * @return the corresponding <code>Date</code> object
+     * @throws IllegalArgumentException If {@code dateString} doesn't match an ISO8601 pattern
+     */
+    public static TimeAndFormat parseDateAndFormat(String dateString) {
```

Comment:
Given that azure-storage-common exposes it's implementation packages to the other azure-storage libraries we should think about moving this into implementation in case we need to do API changes to it in the future.